examples/notebooks/submodel_cracking_DFN_or_SPM.ipynb

```diff
@@ -0,0 +1,264 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Using crack submodels in PyBaMM\n",
+    "In this notebook we show how to use the crack submodel with battery DFN or SPM models. To see all of the models and submodels available in PyBaMM, please take a look at the documentation [here](https://pybamm.readthedocs.io/en/latest/source/models/index.html)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%pip install pybamm -q    # install PyBaMM if it is not installed\n",
+    "import pybamm\n",
+    "import os\n",
+    "import numpy as np\n",
+    "import matplotlib.pyplot as plt\n",
+    "os.chdir(pybamm.__path__[0]+'/..')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Then we load the DFN. When you load a model in PyBaMM it builds by default. Building the model sets all of the model variables and sets up any variables which are coupled between different submodels: this is the process which couples the submodels together and allows one submodel to access variables from another. If you would like to swap out a submodel in an exisitng battery model you need to load it without building it by passing the keyword `build=False`"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#model = pybamm.lithium_ion.SPMe(\n",
+    "model = pybamm.lithium_ion.DFN(\n",
+    "    build=False,options = {\n",
+    "        \"particle\": \"Fickian diffusion\",  \n",
+    "        \"thermal\": \"lumped\", \n",
+    "        \"particle cracking\": \"both\", # other options are \"positive\", \"negative\" or \"none\"\n",
+    "    }\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This collects all of the submodels which make up the DFN, but doesn't build the model. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "param = model.default_parameter_values\n",
+    "\n",
+    "import pandas as pd\n",
+    "mechanics = pd.read_csv(\"pybamm/input/parameters/lithium-ion/mechanicals/lico2_graphite_Ai2020/parameters.csv\", \n",
```

Comment:
Sorry, which default PyBaMM parameter processing command can be used to read those parameters?

Reply:
Not sure if it will work, but could you use `param["mechanics"] = pybamm.parameter_sets.Ai2020`? I think @tinosulzer already commented on a similar issue.

Reply:
No, the set `Ai2020` will give all battery chemistry but here is aimed at only the mechanical part. 
@tinosulzer suggested to merge this file to `lithium_ion_parameters.py`, but will it cause errors in the existing parameter errors without the mechanical part?

Reply:
I think that it will not cause errors as long as the mechanical model is not called with a parameter set that doesn't contain the mechanical models, but @rtimms and @tinosulzer will know better.

Reply:
yeah the parameter only needs to present in the csv if the model uses it. so you should be able to add the mechanical parameters to the anode/cathode for your new parameter set. if people try to use the mechanical model with the other parameter sets it will throw an error saying a parameter is missing

Reply:
I have added all mechanical parameters to `lithium_ion_parameters.py` and my new notebooks work well. But it cannot pass the tests, not surprising, because some parameters cannot be found.

One solution is by setting a condition of adding them. But how to set this condition in `lithium_ion_parameters.py` to let those mechanical parameters loaded, only when the new submodel is being used?

Reply:
Which tests is it failing? Simple solution would be to add the mechanical parameters to all parameters like you did for Marquis2019

Reply:
There was a long list of errors appearing, maybe 35. I can try this simple solution although don't think this is a good way. Eveytime with new parameters, it needs to update all existing parameters.

Reply:
if a model never calls a parameter (the symbol) then it shouldn’t matter that it isn’t is the csv. unless the fact that lithium ion parameters is now a class breaks this? if it does, you could add an if statement in lithium ion parameters so that the mechanical params are only set if the mechanics option is set

Reply:
but i agree, we shouldn’t need to update all the parameter sets every time you add new parameters

Reply:
Can you push the branch with those errors so I can look into it?

Reply:
I haven't pushed it because it cannot pass the tests function. If you don't mind the testing errors, I can do one push and it is done.

The condition I'm using now is `if isinstance(self, pybamm.particle_cracking.BaseCracking):` in `lithium_ion_parameters.py`, but it seems to be always `false`. Any suggestions?

Reply:
yep, fine to push here with the error so we can see what is happening

in `lithium_ion_parameters` you need to check what the option is. e.g. for calculating the surface area per unit volume we do 
```python3
        if self.options["particle shape"] == "spherical":
            self.a_n_dim = 3 * self.epsilon_s_n / self.R_n
            self.a_p_dim = 3 * self.epsilon_s_p / self.R_p
        elif self.options["particle shape"] == "user":
            self.a_n_dim = self.geo.a_n_dim
            self.a_p_dim = self.geo.a_p_dim
```

in the check you have `self` is an instance of the `LithiumIonParameters`

Reply:
since you have to check for this, maybe it _is_ cleaner to keep the mechanical parameters separate? not sure what my opinion is on this. thoughts?

Reply:
Do you mean the parameter values? I don't know what would be a better either. Is there a way to label the parameter values in terms of geometrical, electrochemical, thermal, mechanical... and then you just, let's say, pull the mechanical from Ai2020 but the rest from Chen2020?

In any case, I think we need to rethink how we structure the parameter sets now that we have models beyond pure electrochemical.

Reply:
yeah I think we can probably tackle how we structure parameters as part of a separate issue.

my comment was about having a separate `pybamm.mechanical_parameters`, but just having looked through the PR more closely I think the mechanical parameters (symbol definitions) should live in `LithiumIonParameters` since when you include mechanics you change the definition of the diffusivity (stress assisted). since you want to update parameter definitions in `LithiumIonParameters` it makes sense to define the new ones there too

Reply:
Makes sense. I guess otherwise we would need to do as we did with thermal and redefine all the parameters to depend on the stress, and define a default stress of zero which would make things really messy...

Reply:
Thanks all for your replies. This is a common issue for developing new features, e.g. cracking and Li plating submodels, so I put an overview below and am looking for a solution.

In summary, There are two options: put all new parameters in a separate file; or make them in `LithiumIonParameters` with options to enable therm or not. The latter one seems to make the whole program tightly. But how to make the options for  new models like cracking in `LithiumIonParameters`? I can only see `"particle shape"` there.

Reply:
I *think* you should be able to define them in `LithiumIonParameters` without an option as they should only need to exist in the parameter set if the model needs them. It seems like this isn't working as expected, so I think for this issue we will use the option to enable them, and make a new issue to investigate why it gives an error even if the model doesn't include mechanical effects (and therefore doesn't call those parameters).

To add extra options you'll need to update the options setter with something like 

```python3
    @options.setter
    def options(self, extra_options):
        extra_options = extra_options or {}

        # Default options
        options = {"particle shape": "spherical", "cracking": None}

        # All model options get passed to the parameter class, so we just need
        # to update the options in the default options and ignore the rest
        for name, opt in extra_options.items():
            if name in options:
                options[name] = opt

        # Check the options are valid (this check also happens in 'BaseBatteryModel',
        # but we check here incase the parameter class is instantiated separatly
        # from the model)
        if options["particle shape"] not in ["spherical", "user"]:
            raise pybamm.OptionError(
                "particle shape '{}' not recognised".format(options["particle shape"])
            )
       # ADD CHECK FOR VALID CRACKING OPTIONS HERE

        self._options = options
```